### ISSUE TITLE
Implement hybrid mode support

### ISSUE THREAD

**Issue by rchl** [sublimelsp/LSP-vue#130]
`vue.hybridMode` is mode where `LSP-vue` works together with `LSP-typescript` (with Vue Typescript plugin enabled) where each handles different functionality within the vue file.

`vue.hybridMode` is currently disabled and not recommended because:
 - enabling the Vue Typescript plugin within `LSP-typescript` is a bit cumbersome and I want to provide some helper command for that
 - `LSP-typescript` and `LSP-vue` running together in a vue file fight over which one should handle requests and those don't always end up going to proper servers - this can be fixed by disabling certain capabilities for `LSP-vue` when hybrid mode is enabled

**Comment by rchl** [comment]
A current blocker is related to making the Vue Typescript plugin easy to set up with LSP-typescript.

Setting up Vue Typescript plugin means basically adding a configuration like this in project settings:

```
			"LSP-typescript": {
				"initializationOptions": {
					"plugins": [
						{
							"languages": ["vue"],
							"location": "/Users/rafal/Library/Caches/Sublime Text/Package Storage/LSP-vue/20.18.0/server/node_modules",
							"name": "@vue/typescript-plugin",
						}
					],
				},
			},
```

The problem is the path to the plugin which is not know up-front (it's only known once the LSP-vue is initialized and also it's not ideal that it includes the node version since it means that this setup will break and will have to be updated once node version is updated.

I'd rather define some variable that would provide path to the LSP-vue node_modules but as said before, it's not that straightforward since that path is not known until server is initialized.

I was thinking of maybe copying the vue typescript plugin to a different location but that doesn't sound that great either IMO.

**Comment by rchl** [comment]
For the record (if someone wants to use hybrid mode before it's officially supported), to fix some issues with requests prioritization, it's recommended to also set those for `LSP-vue` when enabling hybrid mode:

```jsonc
{
    // ...
    // Preferred for handling requests in Vue files over LSP-typescript.
    "priority_selector": "text.html.vue source.js, text.html.vue source.ts",
    'initializationOptions': {
        'vue.hybridMode': True,
    },
    'disabled_capabilities': {
        'definitionProvider': True,
        'referencesProvider': True,
        'typeDefinitionProvider': True,
    },
}
```

**Comment by Narretz** [comment]
I activated hybrid mode, and updated the settings, but I don't get diagnostics in my .vue files, neither in the html nor in the script part.

The logs look good:

```
LSP: applying .sublime-project override for LSP-vue
LSP: starting ['C:\\Program Files\\nodejs\\node.EXE', 'C:\\Users\\marti\\AppData\\Local\\Sublime Text\\Package Storage\\LSP-vue\\20.9.0\\server\\node_modules\\@vue\\language-server\\bin\\vue-language-server.js', '--stdio'] in C:\Users\marti\dev
LSP: LSP-vue: Supported execute commands: ['vue.parseSfc', 'vue.detectNameCasing', 'vue.convertTagsToKebabCase', 'vue.convertTagsToPascalCase', 'vue.convertPropsToKebabCase', 'vue.convertPropsToCamelCase']
LSP: LSP-vue: registering capability: didChangeConfigurationProvider
LSP: starting ['C:\\Program Files\\nodejs\\node.EXE', 'C:\\Users\\marti\\AppData\\Local\\Sublime Text\\Package Storage\\LSP-vue\\20.9.0\\server\\node_modules\\@vue\\language-server\\bin\\vue-language-server.js', '--stdio'] in C:\Users\marti\dev
LSP: LSP-vue: Supported execute commands: ['vue.parseSfc', 'vue.detectNameCasing', 'vue.convertTagsToKebabCase', 'vue.convertTagsToPascalCase', 'vue.convertPropsToKebabCase', 'vue.convertPropsToCamelCase']
```

LSP log when hovering over an undefined variable
```
:: [11:53:44.001] --> LSP-vue textDocument/hover (110): {'textDocument': {'uri': 'file:///C:/Users/marti/dev/packages/client/src/components/SideDetail.vue'}, 'position': {'line': 68, 'character': 23}}
:: [11:53:44.005] --> LSP-vue textDocument/codeAction (111): {'textDocument': {'uri': 'file:///C:/Users/marti/dev/packages/client/src/components/SideDetail.vue'}, 'range': {'start': {'line': 68, 'character': 23}, 'end': {'line': 68, 'character': 23}}, 'context': {'diagnostics': [], 'triggerKind': 2}}
:: [11:53:44.071] <<< LSP-vue (110) (duration: 69ms): None
:: [11:53:44.071] <<< LSP-vue (111) (duration: 65ms): []
```

Maybe it's a Windows problem?

**Comment by niksy** [comment]
I think `LSP-typescript` is missing `"selector": "source.js, source.jsx, source.ts, source.tsx, text.html.vue"` in settings, by default it handles only pure JS/TS files, Vue has different scope.

**Comment by rchl** [comment]
I've edited the comments to just have one configuration instead of confusing two different ones.

But yeah, LSP-typescript must be running also.

**Comment by Narretz** [comment]
Thanks!

It works now.

One other thing I had wrong was that I used unix style path for the plugin path, which does not work in Windows. So LSP-typescript started without the plugin and I got garbage diagnostics in .vue files. It would be cool if LSP-vue or LSP-typescript could validate the plugin path somehow.

**Comment by niksy** [comment]
> It would be cool if LSP-vue or LSP-typescript could validate the plugin path somehow. 

This is what @rchl is trying to automate.

**Comment by FichteFoll** [comment]
It read to me as if hybrid mode has been dropped but the LSP client (here: LSP-vue) still needs to make some changes to get cooperation between the vue ls and [vtsls](https://github.com/yioneko/vtsls) going. ts_ls is not supported (yet).

**Comment by niksy** [comment]
Yeah, we still need to implement glue code, but flag for disabling hybrid mode is not needed anymore.

**Comment by niksy** [comment]
How does vtsls relates to Sublime Text implementation? Is that another version of TypeScript language server?

**Comment by niksy** [comment]
To answer myself, vtsls is required since TypeScript language server used in LSP-typescript currently doesn’t support `typescript.tsserverRequest` command. [Issue requesting this feature](https://github.com/typescript-language-server/typescript-language-server/issues/959) is opened so when it does, vtsls shouldn’t be needed for this.

There is LSP client based on vtsls, but I don’t know how well it’s maintained: https://github.com/gxanshu/LSP-vtsls
